assets/js/googlesitekit/widgets/datastore/widgets.js

```diff
@@ -0,0 +1,258 @@
+			registryKey = Object.keys( WidgetComponents ).length + 1;
+			yield {
+				payload: { registryKey },
+				type: SET_WIDGET_COMPONENT_KEY,
+			};
+		}
+
+		// We do this assignment in the action rather than the reducer because we can't send a
+		// payload that includes a React component to the reducer; we'll get an error about
+		// payloads needing to be plain objects.
+		if ( WidgetComponents[ registryKey ] === undefined ) {
+			WidgetComponents[ registryKey ] = {};
+		}
+		if ( WidgetComponents[ registryKey ][ slug ] === undefined ) {
+			WidgetComponents[ registryKey ][ slug ] = settings.component;
+		}
+		const settingsToUse = {
+			priority: 10,
+			...settings,
+		};
+		delete settingsToUse.component;
+
+		yield {
+			payload: { slug, settings: settingsToUse },
+			type: REGISTER_WIDGET,
+		};
+
+		return {};
+	},
+};
+
+export const controls = {};
+
+export const reducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case ASSIGN_WIDGET: {
+			const { slug, areaSlugs } = payload;
+
+			const { areaWidgets } = state;
+			areaSlugs.forEach( ( areaSlug ) => {
+				if ( areaWidgets[ areaSlug ] === undefined ) {
+					areaWidgets[ areaSlug ] = [];
+				}
+
+				if ( ! areaWidgets[ areaSlug ].includes( slug ) ) {
+					areaWidgets[ areaSlug ].push( slug );
+				}
+			} );
+
+			return {
+				...state,
+				areaWidgets,
+			};
+		}
+
+		case REGISTER_WIDGET: {
+			const { slug, settings } = payload;
+
+			if ( state.widgets[ slug ] !== undefined ) {
+				global.console.warn( `Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.` );
+
+				return { ...state };
+			}
+
+			return {
+				...state,
+				widgets: {
+					...state.widgets,
+					[ slug ]: { ...settings, slug },
+				},
+			};
+		}
+
+		case SET_WIDGET_COMPONENT_KEY: {
+			const { registryKey } = payload;
+
+			return {
+				...state,
+				registryKey,
+			};
+		}
+
+		default: {
+			return { ...state };
+		}
+	}
+};
+
+export const resolvers = {
+};
+
+export const selectors = {
+	/**
+	 * Checks if a widget has been registered with a given slug.
+	 *
+	 * Returns `true` if the widget area has been registered.
+	 * Returns `false` if the widget area has NOT been registered.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object}  state Data store's state.
+	 * @param  {string}  slug  Widget's slug.
+	 * @return {boolean}       `true`/`false` based on whether widget has been registered.
+	 */
+	isWidgetRegistered( state, slug ) {
+		const { widgets } = state;
+
+		return widgets[ slug ] !== undefined;
+	},
+
+	/**
+	 * Returns all widgets registered for a given widget area.
+	 *
+	 * Returns an array of all widgets for a given area.
+	 * The widgets are returned in order of their priority, so can be rendered in
+	 * the order provided by the selector.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object} state          Data store's state.
+	 * @param  {string} widgetAreaSlug Widget context to get areas for.
```

Comment:
Same like above here, param and return docs should not be aligned in indentation (only param docs to each other).